llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -1607,24 +1623,80 @@ static LogicalResult commonVerifierPackAndUnPackOp(OpTy packOrUnPack) {
   static_assert(llvm::is_one_of<OpTy, PackOp, UnPackOp>::value,
                 "applies to only pack or unpack operations");
   Operation *op = packOrUnPack.getOperation();
-  int64_t rank = (std::is_same<OpTy, PackOp>::value)
-                     ? packOrUnPack.getInputRank()
-                     : packOrUnPack.getOutputRank();
+  ShapedType unpackedType = (std::is_same<OpTy, PackOp>::value)
+                                ? packOrUnPack.getInputType()
+                                : packOrUnPack.getOutputType();
+  int64_t unpackedRank = unpackedType.getRank();
   SmallVector<int64_t> innerDimsPos =
       extractFromI64ArrayAttr(packOrUnPack.getInnerDimsPos());
   SmallVector<int64_t> outerDimPerm =
       extractFromI64ArrayAttr(packOrUnPack.getOuterDimsPerm());
   // Verify tiles. Make sure each provided tile is non-zero.
-  if (hasZeros(packOrUnPack.getMixedTiles()))
+  SmallVector<OpFoldResult> mixedTiles = packOrUnPack.getMixedTiles();
+  if (hasZeros(mixedTiles))
     return op->emitError("invalid tile factor");
-  if (isInvalid(innerDimsPos, rank))
+  if (isInvalid(innerDimsPos, unpackedRank))
     return op->emitError("invalid inner_dims_pos vector");
-  if (isInvalid(outerDimPerm, rank))
+  if (isInvalid(outerDimPerm, unpackedRank))
     return op->emitError("invalid outer_dims_perm vector");
-  if (packOrUnPack.getMixedTiles().size() != innerDimsPos.size()) {
+  if (mixedTiles.size() != innerDimsPos.size()) {
     return op->emitError(
         "blocking factors must equal the number of dimensions to block");
   }
+
+  // Blocking factors must be less or equal than the input rank, and must
+  // match the number of `dims_pos`.
+  if (mixedTiles.size() > unpackedRank) {
+    return op->emitError(
+        "blocking factors must be less or equal than the input rank");
+  }
+
+  ShapedType packedType = (std::is_same<OpTy, PackOp>::value)
+                              ? packOrUnPack.getOutputType()
+                              : packOrUnPack.getInputType();
+  int64_t packedRank = packedType.getRank();
+  // Require output rank to match input rank + number of blocking factors.
+  if (unpackedRank + mixedTiles.size() != packedRank) {
+    return op->emitError(
+        "packed rank must equal unpacked rank + blocking factors");
+  }
+
+  // Verify result shape is greater than the minimum expected
+  // by the pack operation, and that the output shape
+  // represents full tiles.
+  // Verify result type against inferred type.
+  ShapedType expectedPackedType = PackOp::getPackedType(
+      unpackedType, packOrUnPack.getStaticTiles(), innerDimsPos, outerDimPerm);
+  if (!isSmallerThan(expectedPackedType.getShape(), packedType.getShape())) {
```

Comment:
What use case do you have for providing a bigger tensor? If we want to support this use case, DPS would be the right way. For example, PadOp will allocate a tensor big enough to store the padded tensor.

Reply:
In IREE we might not have all the information to find the exact inner tile size at the point where we are doing the data layout transformation. So the approach taken for now is we "over-pad" to a reasonable value (say 16) that is good enough for all tile sizes you would use. We plan to revisit this later to see if we can avoid over-padding. But this means making the padding being "strict ceil-div" is a bit of an over constraint. So relaxing this to require the output shape to be greater than the minimum required.

Agree that this would work best with DPS. We can maybe make it work without it, but would be much harder....

Reply:
Here DPS in my opinion is the right way.